### DISCUSSION TITLE
collaboration using Visual Studio Code

### DISCUSSION THREAD

**Discussion by bijlevel** [RomanKubiak/ctrlr#232] [Q&A]
I have installed Visual Studio Code on my Mac because it seems promising in collaborating on panels. I could open a file with .panel extension, but then it shows (in essence, I believe) the Utility window of Ctrlr, not the various methods/Lua code. One can download support for a wealth of tools and languages, but before I try them all (I never will of course ;-)) my question: does anyone have experience with Visual Studio and know which tool/language support to use for Ctrlr?

**Comment by Godlike-Productions** [comment]
I use visual studio for lots of stuff.  

For ctrlr my workflow is this.

Create a folder on your Hard Drive where you want the repository to live. Leave it empty.

When you open Visual Studio Code, click on clone repository.  Put the repository URL in the box.  You can find this by clicking on the green code button on the github repo.  It should be something like this.

https://github.com/Godlike-Productions/K2x00-Remote.git

If you are creating a new project, make a new repository on github first and stick a readme file in it.  Don't put anything else in yet.

You may get prompted for your github credentials

You shouldn't need any plugins installed. Visual Studio Code will prompt you if it recognizes the file type (I think LUA is there by default)

Down in the bottom blue bar there is a little sync button. Click that to pull all files from the repository.  

For a new project, save your new panel into your repo folder on your hard drive. For an existing folder, you can open the panel in ctrlr that has just been imported.

In ctrlr, in the LUA editor go to File->Convert to files.  This will export all of the LUA as files.  You need to do this in your new panel.  Opening an existing panel where this has already been done should be OK.  The LUA methods will have little disks instead of the gear icon if they are on disk. New methods always have gear icons.  You'll need to convert to files at the end of your session (this exports all files).  If an old copy of the file exists for any gear methods, they will still save but will be called method2.  This happens for the Built In methods.  I don't sync these if they don't change, and periodically delete them.  They aren't that important, but for your own methods, it's best to keep the latest version and remove old versions from your folder on disk.  If you need to delete the method and copy it from the repo to make a gear icon, and then convert to files.

Do your editing in ctrlr.  You can do it in visual studio, but ctrlr won't automatically update.  If you edit in ctrlr, Visual Studio will automatically update after you save and compile.

When you are ready to commit the change (or send a pull request) click the little fork icon in visual studio.

You will see a list of files that have changed.  Click the refresh icon in the top of Source Control to update.  On all the changed files, if you hover over them a little + will come up. To stage to the repo, click the plus. It will appear up the top. When you have reviewed all changes and staged them. Type a message and then press the tick above.  This will save the changes to your local repo. To sync with github, press the sync in the bottom left in the blue bar.  This will sync with github.

**Reply by bijlevel** [reply]
Thanks for this "manual"! I did ask this on purpose on this forum because that way maybe other members can profit from this software and knowledge of collaborating. My main concern in collaborating is keep track of the various versions that develop when working on the same panel. But it looks like Visual Studio Code does this job well. BTW it is a pity that the Mac version of Ctrlr I'm using doesn't have the option to export the LUA code as files, so I probably have to switch to the Windows version, which does have that option apparently.
